### PR TITLE
Removing the std::pow from the Quadrilateral3D4N jacobian

### LINKED ISSUE #566: TestQuadrilateralJacobianTest is failing
This test is failing in my linux machine with following message 

```
Error: Check failed becuase detJ = inf is not near to MathUtils<double>::GeneralizedDet(jacobian) = 0.0198539 within the tolerance 1e-06
in kratos/tests/geometries/test_non_rectangular_jacobian.cpp:108:virtual void Testing::TestQuadrilateralJacobianTest::TestFunction()

```

@loumalouomega do you think that it is affected by your change in generailize det?

### PR DESCRIPTION
**Pull request by loumalouomega**
Fix #566, I think

### CODE REVIEW

**Comment by loumalouomega** [comment]
Still having the following messages in Vlagrind:

~~~c
==22205==  Uninitialised value was created by a heap allocation
==22205==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22205==    by 0x9127223: __gnu_cxx::new_allocator<double>::allocate(unsigned long, void const*) (new_allocator.h:111)
==22205==    by 0x9123EC1: boost::numeric::ublas::unbounded_array<double, std::allocator<double> >::unbounded_array(unsigned long, std::allocator<double> const&) (storage.hpp:68)
==22205==    by 0x9112D9C: boost::numeric::ublas::matrix<double, boost::numeric::ublas::basic_row_major<unsigned long, long>, boost::numeric::ublas::unbounded_array<double, std::allocator<double> > >::matrix(unsigned long, unsigned long) (matrix.hpp:133)
==22205==    by 0x958D94B: Kratos::Quadrilateral3D4<Kratos::Node<3ul, Kratos::Dof<double> > >::DeterminantOfJacobian(unsigned long, Kratos::GeometryData::IntegrationMethod) const (quadrilateral_3d_4.h:849)
~~~
and 
~~~c
==22205== Conditional jump or move depends on uninitialised value(s)
==22205==    at 0x58B5B48: sqrt (w_sqrt_compat.c:27)
==22205==    by 0x958DD8E: Kratos::Quadrilateral3D4<Kratos::Node<3ul, Kratos::Dof<double> > >::DeterminantOfJacobian(unsigned long, Kratos::GeometryData::IntegrationMethod) const (quadrilateral_3d_4.h:864)
~~~
Corresponding  to: 

imagen

**Comment by pooyan-dadvand** [comment]
It seems that you found an error:
~~~c
==22205== Conditional jump or move depends on uninitialised value(s)
==22205==    at 0x58B5B48: sqrt (w_sqrt_compat.c:27)
==22205==    by 0x958DD8E: Kratos::Quadrilateral3D4<Kratos::Node<3ul, Kratos::Dof<double> > >::DeterminantOfJacobian(unsigned long, Kratos::GeometryData::IntegrationMethod) const (quadrilateral_3d_4.h:864)
~~~

This means that the det_j is has some not initialized value in its calculation (which should be from Jacobian)

**Comment by loumalouomega** [comment]
imagen

It looks OK

**Comment by pooyan-dadvand** [comment]
The value yes, but this due to the fact that the uninitialized value is interpreted as zero. I would suggest you to run the code with watches with valgrind and see which of them is reported by valgrind.

**Comment by loumalouomega** [comment]
It is solved but still a problem:

~~~c
==28953== Conditional jump or move depends on uninitialised value(s)
==28953==    at 0x58B5B48: sqrt (w_sqrt_compat.c:27)
==28953==    by 0x918C257: Kratos::MathUtils<double>::GeneralizedDet(boost::numeric::ublas::matrix<double, boost::numeric::ublas::basic_row_major<unsigned long, long>, boost::numeric::ublas::unbounded_array<double, std::allocator<double> > > const&) (math_utils.h:622)
~~~

**Comment by loumalouomega** [comment]
Solved

**Comment by pooyan-dadvand** [comment]
What was the problem?

**Comment by loumalouomega** [comment]
The jacobian wasn't initialized to zero

**Comment by loumalouomega** [comment]
imagen

File: kratos/geometries/quadrilateral_3d_4.h [hunk 1]
```diff
@@ -839,15 +849,22 @@ template<class TPointType> class Quadrilateral3D4
         IntegrationMethod ThisMethod 
         ) const override
     {
-        Matrix jacobian ( 3, 2 );
+        Matrix jacobian( 3, 2 );
          
-        this->Jacobian( jacobian, IntegrationPointIndex, ThisMethod);
+        Jacobian( jacobian, IntegrationPointIndex, ThisMethod);
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Are you sure you want to use only this class Jacobian?

**Review by pooyan-dadvand — COMMENTED** [review]
I'm not sure if removing this-> is desirable here. 

Also I would not copy the references to matrix values.

**Comment by loumalouomega** [comment]
OK

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi do you have some comments?

**Comment by RiccardoRossi** [comment]
my comment is to use 

        std::pow(number,2)
instead of 
        std::pow(numer,2.0)

the key to performance is to use an integer number instead of a double

**Comment by loumalouomega** [comment]
I removed the std::pow, so there is no need for that

**Comment by pooyan-dadvand** [comment]
@loumalouomega as you wrote it is very verbose but acceptable. But I would use the value and not the references to the value:

```c++
const double j00 = jacobian[pnt](0,0);
```

Which is easier to understand and would give better perspective to the optimizer. (The latter is arguable)

**Comment by loumalouomega** [comment]
@pooyan-dadvand I thought that was the opposite, that with the reference it was more efficient because it didn't create a copy

**Comment by pooyan-dadvand** [comment]
But copying a 64 bit pointer (the reference) is the same as the double value. But having a constant copy may help the optimizer a little more. Nevertheless the difference is not very important in your case.

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

**Pull request merged**